blockmanager_test.go

```diff
@@ -0,0 +1,230 @@
+			cfh := currentCFHeader
+			checkpoints = append(checkpoints, &cfh)
+		}
+
+	}
+
+	// Write all block headers but the genesis, since it is already in the
+	// store.
+	if err = hdrStore.WriteHeaders(blockHeaders[1:]...); err != nil {
+		t.Fatalf("Error writing batch of headers: %s", err)
+	}
+
+	var wg sync.WaitGroup
+
+	// Set up a chain service with a custom query batch method.
+	cs := &ChainService{
+		BlockHeaders: hdrStore,
+	}
+	cs.queryBatch = func(msgs []wire.Message, f func(*ServerPeer,
+		wire.Message, wire.Message) bool, q <-chan struct{},
+		qo ...QueryOption) {
+
+		// Craft response for each message.
+		for _, msg := range msgs {
+			// Only GetCFHeaders expected.
+			q, ok := msg.(*wire.MsgGetCFHeaders)
+			if !ok {
+				t.Fatalf("got unexpected message %T", msg)
+			}
+
+			// The start height must be set to a checkpoint
+			// height+1.
+			if q.StartHeight%wire.CFCheckptInterval != 1 {
+				t.Fatalf("unexpexted start height %v",
+					q.StartHeight)
+			}
+
+			var prevFilterHeader chainhash.Hash
+			switch q.StartHeight {
+
+			// If the start height is 1 the prevFilterHeader is set
+			// to the genesis header.
+			case 1:
+				prevFilterHeader = *genesisFilterHeader
+
+			// Otherwise we use one of the created checkpoints.
+			default:
+				j := q.StartHeight/wire.CFCheckptInterval - 1
+				prevFilterHeader = *checkpoints[j]
+			}
+
+			resp := &wire.MsgCFHeaders{
+				FilterType:       q.FilterType,
+				StopHash:         q.StopHash,
+				PrevFilterHeader: prevFilterHeader,
+			}
+
+			// Keep adding filter hashes until we reach the stop
+			// hash.
+			for h := q.StartHeight; ; h++ {
+				resp.FilterHashes = append(resp.FilterHashes,
+					&cfHeaders[h])
+
+				blockHash := blockHeaders[h].BlockHash()
+				if blockHash == q.StopHash {
+					break
+				}
+			}
+
+			// To ensure the block manager can handle responses
+			// coming potentially out of order, we launch a
+			// goroutine for each, and make it sleep a short while
+			// before calling the response handler.
+			wg.Add(1)
+			go func(query, response wire.Message) {
+				defer wg.Done()
+
+				r := rand.Intn(1000)
+				time.Sleep(time.Duration(r) * time.Millisecond)
+
+				if !f(nil, query, response) {
+					t.Fatalf("got response false")
+				}
+			}(q, resp)
+
+		}
+	}
+
+	// Set up a blockManager with the chain service we defined...
+	bm := blockManager{
+		server: cs,
+		blkHeaderProgressLogger: newBlockProgressLogger(
+			"Processed", "block", log,
+		),
+		fltrHeaderProgessLogger: newBlockProgressLogger(
+			"Verified", "filter header", log,
+		),
+	}
+	bm.newHeadersSignal = sync.NewCond(&bm.newHeadersMtx)
+	bm.newFilterHeadersSignal = sync.NewCond(&bm.newFilterHeadersMtx)
+
+	// ...and call the get checkpointed cf headers method with the
+	// checkpoints we created.
+	bm.getCheckpointedCFHeaders(
+		checkpoints, cfStore, wire.GCSFilterRegular,
+	)
+
+	// Wait for all the responses to be handled.
+	wg.Wait()
```

Comment:
As is, we can't be sure that all those goroutines have actually executed (unless the call to `queryBatch` there is actually sync?) before this call to wait. 

Also in the cases I've seen reported, this happens even with a _single_ peer. In the case of a single peer, the batch execution goroutine will only call one `checkResponse` at a time, so it doesn't seem that this bug arises due to a race.